c++/src/capnp/compat/http-over-capnp.c++

```diff
@@ -416,6 +466,41 @@ public:
     }
   }
 
+  kj::Promise<void> connect(
+      kj::StringPtr host, const kj::HttpHeaders& headers, kj::AsyncIoStream& connection,
+      ConnectResponse& tunnel) override {
+    auto rpcRequest = inner.connectRequest();
+    auto downPipe = kj::newOneWayPipe();
+    rpcRequest.setHost(host);
+    rpcRequest.setDown(factory.streamFactory.kjToCapnp(kj::mv(downPipe.out)));
+
+    auto builder = capnp::Request<
+        capnp::HttpService::ConnectParams,
+        capnp::HttpService::ConnectResults>::Builder(rpcRequest);
+    rpcRequest.adoptHeaders(factory.headersToCapnp(headers,
+        Orphanage::getForMessageContaining(builder)));
+    rpcRequest.setContext(
+        kj::heap<ConnectClientRequestContextImpl>(factory, tunnel));
+    RemotePromise<capnp::HttpService::ConnectResults> pipeline = rpcRequest.send();
+
+    auto promises = kj::heapArrayBuilder<kj::Promise<void>>(2);
+    // We read from `downPipe` (the other side writes into it.)
+    auto downPumpTask = downPipe.in->pumpTo(connection).then([&connection](uint64_t) {
+      connection.shutdownWrite();
+    }).attach(kj::mv(downPipe.in)).eagerlyEvaluate(nullptr);
```

Comment:
Since you're no longer joining promises, and you've passed `nullptr` to `eagerlyEvaluate()`, any errors from `downPumpTask` won't be reported anywhere.

The reason `eagerlyEvaluate()` takes a parameter is actually to remind you to handle errors. The parameter is a callback function for the error case, just like `catch_()` would take. It's only appropriate to pass `nullptr` in cases where you are actually waiting on the promise later and so the errors will propagate.

Admittedly, it's not entirely clear what we should do with the error here. When there's truly nothing better to do, I generally recommend logging with `KJ_LOG(ERROR)`.

But I think we probably actually want fail-fast behavior. That is, if either the up-stream or the down-stream pump throws an exception, we want to cancel out the whole operation, propagating that exception.

Now, naively, that suggests we should just `exclusiveJoin()` all the promises together. With `exclusiveJoin()`, if any promise rejects, the others are canceled. But there's a problem, which is that we actually _don't_ want to proactively cancel the operation in the case that the pump in one direction completes successfully. One trick we can do is to make the pump promises return `kj::NEVER_DONE` at the end of the success path:

```
auto downPumpTask = downPipe.in->pumpTo(connection)
    .then([&connection, down = kj::mv(downPipe.in)](uint64_t) {
  connection.shutdownWrite();
  return kj::NEVER_DONE;
});
```

(Aside: Note that instead of using `.attach()` to attach `downPipe.in`, I captured it into the `.then()` continuation. This has basically the same effect, but avoids an extra promise node allocation, and also in ensures that the object is dropped before the NEVER_DONE wait.)

With this pattern, now we can use `exclusiveJoin()` to combine everything at the end of the method:

```
return pipeline.ignoreResult()
    .exclusiveJoin(kj::mv(downPumpTask))
    .exclusiveJoin(kj::mv(upPumpTask));
```